Creation of area and iteration

dennismdejong opened the pull request:
Creation of area and iteration as requested in #386

## All Submissions:

* [ x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ x] I have updated the documentation accordingly.
* [ x] I have added tests to cover my changes.
* [x ] All new and existing tests passed.
* [ x] My code follows the code style of this project.
* [ x] I ran lint checks locally prior to submission.
* [ x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #386 #394

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [ x] No

<!-- If this introduces a change to these files, please elaborate on why -->

## Does this introduce a breaking change?

- [ ] Yes
- [ x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

dennismdejong committed:
Creation of area and iteration

Creation of area and iteration as requested in #386

dennismdejong committed:
test: Missed the acceptance tests

dennismdejong committed:
fix: Failing linters

dennismdejong committed:
feat: Creating the basis resources

dennismdejong committed:
fixing terraformfmt

magodo commented on website/docs/r/area.html.markdown:
```diff
@@ -0,0 +1,51 @@
+
+---
+
+* `path` - (Optional) Area. Changing this forces a new Project Area to be created.
```
Same here

magodo commented on website/docs/r/area.html.markdown:
```diff
@@ -0,0 +1,51 @@
+
+* `has_children` - Area.
+
+* `node_id` - The ID of the Area.
```
Same here

magodo commented on azuredevops/internal/service/workitemtracking/utils/classification.go:
```diff
@@ -76,6 +76,56 @@ func CreateClassificationNodeSchema(outer map[string]*schema.Schema) map[string]
+					},
+					"finish_date": {
+						Type:     schema.TypeString,
+						Optional: true,
```
Can we add an validation on the format of them?

magodo commented on azuredevops/internal/service/workitemtracking/utils/classification.go:
```diff
@@ -120,6 +171,90 @@ func ReadClassificationNode(clients *client.AggregatedClient, d *schema.Resource
+		Name: converter.String(d.Get("name").(string)),
+	}
+
+	if attributesData, ok := d.GetOk("attributes"); ok {
```
For the purpose of maintainability, can we avoid reusing a single function for the Create/Update among the two resource types? E.g. here the `attributes` only exists for the iteration resource. Though you can combine them together with this, it makes the code hard to maintain, or even hardly to migrate to `terraform-plugin-framework`, where the way to retieve an attribute is strictly typed.

magodo commented on azuredevops/internal/service/workitemtracking/resource_area.go:
```diff
@@ -0,0 +1,40 @@
+
+func resourceReadArea(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	return utils.ReadClassificationNode(clients, d, workitemtracking.TreeStructureGroupValues.Areas)
```
We shall not couple this resource to the data source's Get implementation.

magodo commented on azuredevops/internal/service/workitemtracking/utils/classification.go:
```diff
@@ -120,6 +171,90 @@ func ReadClassificationNode(clients *client.AggregatedClient, d *schema.Resource
+		d.SetId(workItem.Identifier.String())
+	} else if workItem.Id != nil {
+		d.SetId(fmt.Sprintf("%d", *workItem.Id))
+	}
```
Since the [read API](https://learn.microsoft.com/en-us/rest/api/azure/devops/wit/classification-nodes/get-classification-nodes?view=azure-devops-rest-7.2&tabs=HTTP) uses the `ids` to read a specific classification resource, shall we make sure to only set `workItem.Id` as its id?

If the `Identifier` is useful in other context, we shall expose (at least) the `identifier` as another Computed attribute.

magodo commented on azuredevops/internal/acceptancetests/resource_area_test.go:
```diff
@@ -0,0 +1,128 @@
+  name       = "%s"
+  path       = "\\"
+}`, projectName, areaName)
+}
```
Can we add more tests to cover the non-root path area?

magodo reviewed (changes requested):
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

dennismdejong committed:
Merge remote-tracking branch 'upstream/main'

dennismdejong committed:
Merge remote-tracking branch 'upstream/main'

dennismdejong committed:
Refactor workitemtracking iteration and area based on PR review

This commit addresses feedback from PR #1452:
- Removes accidentally committed binary and unrelated docs
- Renames resources to azuredevops_workitemtracking_iteration and azuredevops_workitemtracking_area
- Decouples logic in classification.go and moves Create/Update/Delete directly into resources
- Uses workItem.Id as the tf resource id instead of path
- Adds identifier as a computed attribute
- Adds format validation for names
- Adds non-root path acceptance tests

dennismdejong committed:
Resolve merge conflicts with main

dennismdejong committed:
Update go module dependencies to fix testing framework conflict

dennismdejong committed:
Migrate testing framework imports from sdk/v2 to plugin-testing

dennismdejong committed:
Revert testing framework upgrade to v1.15 to maintain Go 1.24 compatibility

dennismdejong committed:
Fix formatting issues identified by golangci-lint in resource_audit_stream.go

dennismdejong committed:
Fix error handling in resourceAuditStreamUpdate function

dennismdejong committed:
Remove azuredevops/internal/service/audit and related client code

dennismdejong committed:
Complete removal of audit service (resolving delete/modify conflicts)

dennismdejong committed:
Update vendored dependencies

dennismdejong committed:
Update vendored dependencies

dennismdejong committed:
Merge remote-tracking branch 'origin/main' into issue_386

dennismdejong committed:
Merge branch 'main' of https://github.com/microsoft/terraform-provider-azuredevops into issue_386 and resolve conflicts

jackmtpt commented:
@nmangue did you manage to get the PR updated / into a mergeable state?

magodo commented:
Wondering why this PR introduces some other new resources. We shall keep one PR to one single resource.

nmangue commented:
Hello @jackmtpt, @dennismdejong made changes to have a mergeable PR and has requested a review. Could you review them please ?

jackmtpt commented:
@nmangue definitely not, I don't know go well enough to review a PR unfortunately.

dennismdejong committed:
Merge upstream/main and resolve conflicts

- Accept upstream version of resource_area.go
- Add azuredevops_workitemtracking_iteration resource and data source
- Keep upstream's azuredevops_area registration
- Remove accidentally merged aquaplatform files
- Remove AGENTS.md

dennismdejong committed:
Fix CI issues: restore data_area_test.go, revert vendor/ to upstream, fix provider names
